heartbeat/security.go

```diff
@@ -0,0 +1,236 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// TODO: Support other architectures (e.g. arm)
+// +build linux
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"runtime"
+	"strconv"
+	"syscall"
+
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+
+	"github.com/elastic/beats/v7/libbeat/common/seccomp"
+)
+
+func init() {
+	// Here we set a bunch of linux specific security stuff.
+	// In the context of a container, where users frequently run as root, we follow BEAT_SETUID_AS to setuid/gid
+	// and add capabilities to make this actually run as a regular user. This also helps node in synthetics, which
+	// does not want to run as root. It's also just generally more secure.
+	if localUserName := os.Getenv("BEAT_SETUID_AS"); localUserName != "" && syscall.Geteuid() == 0 {
+		err := changeUser(localUserName)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// Attempt to set capabilities before we setup seccomp rules
+	// Note that we discard any errors because they are not actionable.
+	// The beat should use `getcap` at a later point to examine available capabilities
+	// rather than relying on errors from `setcap`
+	setCapabilities()
+
+	switch runtime.GOARCH {
+	case "amd64", "386":
+		err := setSeccompRules()
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func changeUser(localUserName string) error {
+	localUser, err := user.Lookup(localUserName)
+	if err != nil {
+		return fmt.Errorf("could not lookup '%s': %w", localUser, err)
+	}
+	localUserUid, err := strconv.Atoi(localUser.Uid)
+	if err != nil {
+		return fmt.Errorf("could not parse UID '%s' as int: %w", localUser.Uid, err)
+	}
+	localUserGid, err := strconv.Atoi(localUser.Gid)
+	if err != nil {
+		return fmt.Errorf("could not parse GID '%s' as int: %w", localUser.Uid, err)
+	}
+	// We include the root group because the docker image contains many directories (data,logs)
+	// that are owned by root:root with 0775 perms. The heartbeat user is in both groups
+	// in the container, but we need to repeat that here.
+	err = syscall.Setgroups([]int{localUserGid, 0})
+	if err != nil {
+		return fmt.Errorf("could not prsetgroups: %w", err)
+	}
+
+	// Set the main group as localUserUid so new files created are owned by the user's group
+	err = syscall.Setgid(localUserGid)
+	if err != nil {
+		return fmt.Errorf("could not set gid to %d: %w", localUserGid, err)
+	}
+
+	// Note this is not the regular SetUID! Look at the 'cap' package docs for it, it preserves
+	// capabilities post-SetUID, which we use to lock things down immediately
+	err = cap.SetUID(localUserUid)
+	if err != nil {
+		return fmt.Errorf("could not setuid to %d: %w", localUserUid, err)
```

Comment:
```suggestion
		return fmt.Errorf("could not set uid to %d: %w", localUserUid, err)
```

Reply:
I know this seems incorrect, but the syscall is `setuid` not `set uid`. I guess both are correct, but I feel `setuid` is more specific (it's a kernel syscall).

Reply:
Gotcha 👍🏽